Use array for device handles

reynir opened the pull request:
A device handle is a 64 bit number with exactly one bit set. Thus we can have up to 63 devices. We can then use an array instead of a map for the `Lwt_condition.t`s that we use to wait and signal when a device is ready. We also don't "lazily" create `Lwt_condition.t`s anymore, and instead the 64 conditions are created at module initialization (allocation of one block each).

There is an assumption that the `int64`s passed are valid handles. If they are not you may now wait for a valid handle instead of an invalid handle that will never resolve. We could try add a check for that (should it be a runtime error? do we wait indefinitely then?)

I have tested it with just one device. I'm not very confident in bit fiddling, and I'm not sure how portable `__builtin_clzll` is (and we should maybe use `__builtin_ctzll` instead). We should also be able to use an array of length 63 instead of 64, but then I have to think even harder.

I think this won't give us massive performance boost, but since it's at the core of all IO it may be worth it still.

reynir committed:
Use an array for device handles

reynir committed:
Use __builtin_clzll over stdc_first_leading_one_ull

We don't seem to have stdbit.h on solo5. Not sure how portable this is.

reynir committed:
Rename binding

hannesm commented:
The question I have while this API gets revised: could we use an (immediate) int instead of of (boxed) int64?

And: where's the limit of 64 devices coming from?

reynir commented:
I couldn't find one place where it's nicely documented, but you can already tell from this snippet:

https://github.com/mirage/mirage-solo5/blob/be95f5a5b1af6d19ce60c1ab2475d27c007f824a/lib/main.ml#L58-L67

The ready set is a 64 bit integer used as a bit vector. A bit set means that device is ready for IO.

We could likely use an immediate int if we do some massaging of handles. `0x8000_0000_0000_0000L` (Int64.min_int) is a valid handle, but since `0L` is not a valid handle (it's reserved, and would not be usable in the bit vector) we can shift one bit and store it in an immediate int. I can give that a try.

hannesm commented:
I was thinking that we could as well just cut it down from 64 to 63 network devices.

hannesm commented:
also please note I've no clue whether this makes any difference, I was just curious.

dinosaure commented:
> And: where's the limit of 64 devices coming from?

It comes from Solo5:
```c
/*
 * Type for sets of up to 64 I/O handles.
 */
typedef uint64_t solo5_handle_set_t;
```

From my perspective, we really should do a benchmark to see the impact of such a change. It can be "easy" to insert a bug in that part. We should make sure it's worth it.

reynir commented:
I was not feeling so confident about the number 63.

We have `MFT_MAX_ENTRIES` defined to 64:
https://github.com/Solo5/solo5/blob/ff63477507a032583e2e2a91dc93eed5b27103cd/include/mft_abi.h#L87-L90
And in `mft_validate()` we require that the number of `entries <= MFT_MAX_ENTRIES` and that entry 0 (first one) is `MFT_RESERVED_FIRST` (not a device). So entries with index 1 through 63 are valid devices.
https://github.com/Solo5/solo5/blob/ff63477507a032583e2e2a91dc93eed5b27103cd/tenders/common/mft.c#L32-L81

With that in mind I'm not sure `0x8000_0000_0000_0000` is a valid device handle after all.

It's a good point about benchmarks. I will try to write a test and a benchmark.

reynir commented:
I wrote a benchmark which suggests there is a small improvement, but it is probably too noisy to draw conclusions from.

Source available here: https://gist.github.com/reynir/b8ff35fa7de0dc880522ff83a2bde04d

Edit: I found I was on the `fix-test` branch so these numbers are for more or less the same code. Testing against the right branch I don't see anything interesting.

<details><summary>performance run output</summary>

```shell
reynir@solsort(vim is open):~/workspace/mirage-solo5$ yes | solo5-hvt --net:stdin=@0 _build/solo5/test64/perf.exe 
            |      ___|
  __|  _ \  |  _ \ __ \
\__ \ (   | | (   |  ) |
____/\___/ _|\___/____/
Solo5: Bindings version v0.9.1
Solo5: Memory map: 512 MB addressable:
Solo5:   reserved @ (0x0 - 0xfffff)
Solo5:       text @ (0x100000 - 0x1f0fff)
Solo5:     rodata @ (0x1f1000 - 0x22dfff)
Solo5:       data @ (0x22e000 - 0x2c5fff)
Solo5:       heap >= 0x2c6000 < stack < 0x20000000
Read 1000000 bytes in 1.042s
Solo5: solo5_exit(0) called
reynir@solsort(vim is open):~/workspace/mirage-solo5$ yes | solo5-hvt --net:stdin=@0 _build/solo5/test64/perf.exe 
            |      ___|
  __|  _ \  |  _ \ __ \
\__ \ (   | | (   |  ) |
____/\___/ _|\___/____/
Solo5: Bindings version v0.9.1
Solo5: Memory map: 512 MB addressable:
Solo5:   reserved @ (0x0 - 0xfffff)
Solo5:       text @ (0x100000 - 0x1f0fff)
Solo5:     rodata @ (0x1f1000 - 0x22dfff)
Solo5:       data @ (0x22e000 - 0x2c5fff)
Solo5:       heap >= 0x2c6000 < stack < 0x20000000
Read 1000000 bytes in 1.039s
Solo5: solo5_exit(0) called
reynir@solsort(vim is open):~/workspace/mirage-solo5$ yes | solo5-hvt --net:stdin=@0 _build/solo5/test64/perf.exe 
            |      ___|
  __|  _ \  |  _ \ __ \
\__ \ (   | | (   |  ) |
____/\___/ _|\___/____/
Solo5: Bindings version v0.9.1
Solo5: Memory map: 512 MB addressable:
Solo5:   reserved @ (0x0 - 0xfffff)
Solo5:       text @ (0x100000 - 0x1f0fff)
Solo5:     rodata @ (0x1f1000 - 0x22dfff)
Solo5:       data @ (0x22e000 - 0x2c5fff)
Solo5:       heap >= 0x2c6000 < stack < 0x20000000
Read 1000000 bytes in 1.063s
Solo5: solo5_exit(0) called
reynir@solsort(vim is open):~/workspace/mirage-solo5$ git checkout main 
Switched to branch 'main'
Your branch is up to date with 'origin/main'.
reynir@solsort(vim is open):~/workspace/mirage-solo5$ dune b
reynir@solsort(vim is open):~/workspace/mirage-solo5$ yes | solo5-hvt --net:stdin=@0 _build/solo5/test64/perf.exe 
            |      ___|
  __|  _ \  |  _ \ __ \
\__ \ (   | | (   |  ) |
____/\___/ _|\___/____/
Solo5: Bindings version v0.9.1
Solo5: Memory map: 512 MB addressable:
Solo5:   reserved @ (0x0 - 0xfffff)
Solo5:       text @ (0x100000 - 0x1f0fff)
Solo5:     rodata @ (0x1f1000 - 0x22dfff)
Solo5:       data @ (0x22e000 - 0x2c5fff)
Solo5:       heap >= 0x2c6000 < stack < 0x20000000
Read 1000000 bytes in 992.674ms
Solo5: solo5_exit(0) called
reynir@solsort(vim is open):~/workspace/mirage-solo5$ yes | solo5-hvt --net:stdin=@0 _build/solo5/test64/perf.exe 
            |      ___|
  __|  _ \  |  _ \ __ \
\__ \ (   | | (   |  ) |
____/\___/ _|\___/____/
Solo5: Bindings version v0.9.1
Solo5: Memory map: 512 MB addressable:
Solo5:   reserved @ (0x0 - 0xfffff)
Solo5:       text @ (0x100000 - 0x1f0fff)
Solo5:     rodata @ (0x1f1000 - 0x22dfff)
Solo5:       data @ (0x22e000 - 0x2c5fff)
Solo5:       heap >= 0x2c6000 < stack < 0x20000000
Read 1000000 bytes in 1.032s
Solo5: solo5_exit(0) called
reynir@solsort(vim is open):~/workspace/mirage-solo5$ yes | solo5-hvt --net:stdin=@0 _build/solo5/test64/perf.exe 
            |      ___|
  __|  _ \  |  _ \ __ \
\__ \ (   | | (   |  ) |
____/\___/ _|\___/____/
Solo5: Bindings version v0.9.1
Solo5: Memory map: 512 MB addressable:
Solo5:   reserved @ (0x0 - 0xfffff)
Solo5:       text @ (0x100000 - 0x1f0fff)
Solo5:     rodata @ (0x1f1000 - 0x22dfff)
Solo5:       data @ (0x22e000 - 0x2c5fff)
Solo5:       heap >= 0x2c6000 < stack < 0x20000000
Read 1000000 bytes in 1.022s
Solo5: solo5_exit(0) called
reynir@solsort(vim is open):~/workspace/mirage-solo5$ yes | solo5-hvt --net:stdin=@0 _build/solo5/test64/perf.exe 
            |      ___|
  __|  _ \  |  _ \ __ \
\__ \ (   | | (   |  ) |
____/\___/ _|\___/____/
Solo5: Bindings version v0.9.1
Solo5: Memory map: 512 MB addressable:
Solo5:   reserved @ (0x0 - 0xfffff)
Solo5:       text @ (0x100000 - 0x1f0fff)
Solo5:     rodata @ (0x1f1000 - 0x22dfff)
Solo5:       data @ (0x22e000 - 0x2c5fff)
Solo5:       heap >= 0x2c6000 < stack < 0x20000000
Read 1000000 bytes in 1.003s
Solo5: solo5_exit(0) called
```

</details>

reynir commented:
I realized I was benchmarking the wrong thing: with that setup I/O is probably always available and the `wait_for_work_on_handle` is not exercised. So I rewrote it to exercise that function. The timings for read are not interesting as we only use one device (so it's more an exercise in lwt). Earlier I also added net devices stdout and stderr just so the map was not as shallow. The results are again too noisy to conclude anything.

<details>
<summary>Results</summary>

```shell
reynir@bark:~$ for i in $(seq 10); do { yes; } | sudo solo5-hvt --net:stdin=@0 --net:stdout=@1 --net:stderr=@2 perf-main --solo5:quiet; done
Wait 379.760ms
Wait 407.891ms
Wait 380.064ms
Wait 381.367ms
Wait 405.189ms
Wait 398.955ms
Wait 378.830ms
Wait 379.852ms
Wait 378.255ms
Wait 380.373ms
reynir@bark:~$ for i in $(seq 10); do { yes; } | sudo solo5-hvt --net:stdin=@0 --net:stdout=@1 --net:stderr=@2 perf-flo64 --solo5:quiet; done
Wait 381.342ms
Wait 384.789ms
Wait 401.174ms
Wait 378.556ms
Wait 378.072ms
Wait 384.239ms
Wait 378.135ms
Wait 379.829ms
Wait 379.270ms
Wait 377.929ms
```

</details>

<details>
<summary>

`perf.ml`

</summary>

```OCaml
let app () =
  let open Lwt.Syntax in
  let* netif = Netif.connect "stdin" in
  let iterations = 500_000 in
  let start = Mirage_mtime.elapsed_ns () in
  let reads =
    List.init iterations
      (fun _ ->
         Solo5_os.Main.wait_for_work_on_handle netif.Netif.handle)
  in
  let mid = Mirage_mtime.elapsed_ns () in
  let* _reads = Lwt_list.map_p (Fun.id) reads in
  let _fin = Mirage_mtime.elapsed_ns () in
  Format.printf "Wait %a\n%!"
    Duration.pp (Int64.sub mid start);
  (* Format.printf "Read %a\n%!"
    Duration.pp (Int64.sub fin mid); *)
  Lwt.return_unit

let () = Solo5_os.Main.run (app ())
```

</details>

reynir committed:
Format

reynir commented:
I will put this on hold. It's challenging to benchmark and I struggle with getting tests running in CI. And this only impacts I/O when there are no incoming packets waiting, so likely this is not a hot path.
